plutus-ir/src/Language/PlutusIR.hs

```diff
@@ -0,0 +1,138 @@
+{-# LANGUAGE FlexibleContexts      #-}
+{-# LANGUAGE FlexibleInstances     #-}
+{-# LANGUAGE LambdaCase            #-}
+{-# LANGUAGE MultiParamTypeClasses #-}
+{-# LANGUAGE OverloadedStrings     #-}
+module Language.PlutusIR (
+    TyName (..),
+    Name (..),
+    Kind (..),
+    Type (..),
+    VarDecl (..),
+    TyVarDecl (..),
+    Datatype (..),
+    Recursivity (..),
+    Binding (..),
+    Term (..),
+    prettyDef,
+    embedIntoIR
+    ) where
+
+import           PlutusPrelude
+
+import           Language.PlutusCore        (Kind, Name, TyName, Type)
+import qualified Language.PlutusCore        as PLC
+import qualified Language.PlutusCore.Pretty as PLC
+
+import           GHC.Generics               (Generic)
+
+-- Variable declarations
+
+data VarDecl tyname name a = VarDecl a (name a) (Type tyname a)
+    deriving (Functor, Show, Eq, Generic)
+data TyVarDecl tyname a = TyVarDecl a (tyname a) (Kind a)
+    deriving (Functor, Show, Eq, Generic)
+
+-- Datatypes
+
+data Datatype tyname name a = Datatype a (TyVarDecl tyname a) [TyVarDecl tyname a] (name a) [VarDecl tyname name a]
+    deriving (Functor, Show, Eq, Generic)
+
+-- Bindings
+
+data Recursivity = NonRec | Rec
+    deriving (Show, Eq, Generic)
+data Binding tyname name a = TermBind a (VarDecl tyname name a) (Term tyname name a)
+                           | TypeBind a (TyVarDecl tyname a) (Type tyname a)
+                           | DatatypeBind a (Datatype tyname name a)
+    deriving (Functor, Show, Eq, Generic)
+
+-- Terms
+
+{- Note [Declarations in Plutus Core]
+In Plutus Core declarations are usually *flattened*, i.e. `(lam x ty t)`, whereas
+in Plutus IR they are *reified*, i.e. `(lam (vardecl x ty) t)`.
+
+Plutus IR needs reified declarations to make it easier to parse *lists* of declarations,
+which occur in a few places.
+
+However, we also include all of Plutus Core's term syntax in our term syntax (and we also use
+its types). We therefore have somewhat inconsistent syntax since declarations in Plutus
+Core terms will be flattened. This makes the embedding obvious and makes life easier
+if we were ever to use the Plutus Core AST "for real".
```

Comment:
Why do we need to use PLC in PLC IR? My default would be not to bother with this. And even if I like to do this, I'd just make a separate constructor

```haskell
plcTerm :: Plc.Term tyname name a -> Term tyname name a
```

Why do we need to interleave PLC and PLC IR?

Reply:
It *is* intended to be a superset of PLC. As for why we want to interleave them, not interleaving them essentially means all let-bindings must be at the top level, but local let-bindings are also useful, and I'd like to use PIR lets for them.

Reply:
> It is intended to be a superset of PLC.

Do we have the motivation for this somewhere?

> As for why we want to interleave them, not interleaving them essentially means all let-bindings must be at the top level, but local let-bindings are also useful, and I'd like to use PIR lets for them.

Thanks. That would be a helpful comment in the code.

Reply:
I think mostly just simplicity, and it makes things like adding typechecking judgements or similar easier.

In the long run I hope we can actually figure out a way to make it an "extension" of PLC, which would be very nice (e.g. maybe we can even reuse most of the typechecker).